aequilibrae/paths/route_choice.pyx

```diff
@@ -0,0 +1,253 @@
+        self.lon_view = graph.lonlat_index.lon.values
+        self.predecessors_view = np.empty(graph.compact_num_nodes + 1, dtype=np.int64)
+        self.ids_graph_view = graph.compact_graph.id.values
+        self.conn_view = np.empty(graph.compact_num_nodes + 1, dtype=np.int64)
+
+    def __dealloc__(self):
+        """
+        C level deallocation. For freeing memeory allocated by this object. *Must* have GIL, `self` may be in a
+        partially deallocated state already.
+        """
+        pass
+
+    def run(self, long origin, long destination, unsigned int max_routes=0, unsigned int max_depth=0):
+        cdef:
+            long origin_index = self.nodes_to_indices_view[origin]
+            long dest_index = self.nodes_to_indices_view[destination]
+            double [:] scratch_cost = np.empty(self.cost_view.shape[0])  # allocation of new memory view required gil
+            unordered_set[vector[long long] *] *results
+            unordered_map[unordered_set[long long] *, vector[long long] *].const_iterator results_iter
+        with nogil:
+            results = RouteChoice.generate_route_set(self, origin_index, dest_index, max_routes, max_depth, scratch_cost)
+
+        res = []
+        for x in deref(results):
+            res.append(np.array(deref(x)))
+            del x
+
+        return res
+
+    cdef void path_find(RouteChoice self, long origin_index, long dest_index, double [:] scratch_cost) noexcept nogil:
+        path_finding_a_star(
+            origin_index,
+            dest_index,
+            scratch_cost,
+            self.b_nodes_view,
+            self.graph_fs_view,
+            self.nodes_to_indices_view,
+            self.lat_view,
+            self.lon_view,
+            self.predecessors_view,
+            self.ids_graph_view,
+            self.conn_view,
+            EQUIRECTANGULAR  # FIXME: enum import failing due to redefinition
+        )
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.embedsignature(True)
+    @cython.initializedcheck(False)
+    cdef unordered_set[vector[long long] *] *generate_route_set(RouteChoice self, long origin_index, long dest_index, unsigned int max_routes, unsigned int max_depth, double [:] scratch_cost) noexcept nogil:
+        """Main method for route set generation"""
+        cdef:
+            unordered_map[unordered_set[long long] *, vector[long long] *] *working_set
+            unordered_map[unordered_set[long long] *, vector[long long] *].const_iterator set_iter
+            vector[unordered_set[long long] *] queue
+            vector[unordered_set[long long] *] next_queue
+            unordered_set[long long] *banned
+            unordered_set[long long] *new_banned
+            unordered_set[vector[long long] *] *route_set
+            vector[long long] *vec
+            pair[unordered_set[long long] *, vector[long long] *] *pair_tmp
+            pair[unordered_set[long long] *, vector[long long] *] x
```

Comment:
All of these maps/sets of pointers make no sense. A pointer will be unique for each object we construct. Looks like I might have to implement my own hash functions after all, or custom data structure